docs/cloud-integration.md

```diff
@@ -0,0 +1,512 @@
+[Amazon's own filesystem client](https://aws.amazon.com/premiumsupport/knowledge-center/emr-file-system-s3/),
+one that is closed-source.
+
+As EMR also maps `s3n://` to the same filesystem, using URLs with the `s3n://` schema avoids
+some confusion. Bear in mind, however, that Amazon's S3 client library is not the Apache one:
+only Amazon can field bug reports related to it.
+
+To work with this data outside of EMR itself, use `s3a://` or `s3n://` instead.
+
+
+#### <a name="asf_s3"></a>Obsolete: Apache Hadoop's S3 client, `s3://`
+
+Apache's own Hadoop releases (i.e not EMR), uses URL `s3://` to refer to a
+deprecated inode-based filesystem implemented on top of S3.
+This filesystem is obsolete, deprecated and has been dropped from Hadoop 3.x.
+
+*Important: * Do not use `s3://` URLs with Apache Spark except on Amazon EMR*
+It is not the same as the Amazon EMR one and incompatible with all other applications.
+
+
+### <a name="working_with_azure"></a>Working with Microsoft Azure Storage
+
+Azure support comes with the [`wasb` filesystem client](https://hadoop.apache.org/docs/stable2/hadoop-azure/index.html).
+
+The Apache implementation is that used by Microsoft in Azure itself: it can be used
+to access data in Azure as well as remotely. The object store itself is *consistent*, and
+can be reliably used as the destination of queries.
+
+
+### <a name="working_with_swift"></a>Working with OpenStack Swift
+
+
+The OpenStack [`swift://` filesystem client](https://hadoop.apache.org/docs/stable2/hadoop-openstack/index.html)
+works with Swift object stores in private OpenStack installations, public installations
+including Rackspace Cloud and IBM Softlayer.
+
+### <a name="working_with_google_cloud_storage"></a>Working with Google Cloud Storage
+
+[Google Cloud Storage](https://cloud.google.com/storage) is supported via Google's own
+[GCS filesystem client](https://cloud.google.com/hadoop/google-cloud-storage-connector).
+
+
+For use outside of Google cloud, `gcs-connector.jar` must be be manually downloaded then added
+to `$SPARK_HOME/jars`.
+
+
+## <a name="cloud_stores_are_not_filesystems"></a>Important: Cloud Object Stores are Not Real Filesystems
+
+Object stores are not filesystems: they are not a hierarchical tree of directories and files.
+
+The Hadoop filesystem APIs offer a filesystem API to the object stores, but underneath
+they are still object stores, [and the difference is significant](http://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/filesystem/introduction.html)
+
+While object stores can be used as the source and store
+for persistent data, they cannot be used as a direct replacement for a cluster-wide filesystem such as HDFS.
+This is important to know, as the fact they are accessed with the same APIs can be misleading.
+
+### Directory Operations May be Slow and Non-atomic
+
+Directory rename and delete may be performed as a series of operations. Specifically, recursive
```

Comment:
good point. I'll call that out up front

1. Reading and writing data can be slower than expected.
1. Some directory structures may be very inefficient to scan during query split calculation.
1. The output of saved RDD may not be immediately visible to a follow-on query.
1. The internal mechanism by which Spark commits work when saving an RDD is potentially
both slow and unreliable.

That's essentially it on day to day use.